components/tikv_util/src/file.rs

```diff
@@ -134,6 +137,232 @@ impl<R: Read> Read for Sha256Reader<R> {
+                pos += read;
+                remains -= read;
+                if read < allowed {
+                    break;
+                }
+            }
+            Ok(pos)
+        } else {
+            self.inner.read(buf)
+        }
+    }
+}
+
+impl Write for InstrumentedFile {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if let Some(limiter) = &mut self.limiter {
+            let mut remains = buf.len();
+            let mut pos = 0;
+            while remains > 0 {
+                let allowed = limiter.request(get_io_type(), IOOp::Write, remains);
+                let written = self.inner.write(&buf[pos..pos + allowed])?;
+                pos += written;
+                remains -= written;
+                if written < allowed {
+                    break;
+                }
+            }
+            Ok(pos)
+        } else {
+            self.inner.write(buf)
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+}
+
+/// Indicates how large a buffer to pre-allocate before reading the entire file.
+fn initial_buffer_size(file: &InstrumentedFile) -> usize {
+    // Allocate one extra byte so the buffer doesn't need to grow before the
+    // final `read` call at the end of the file.  Don't worry about `usize`
+    // overflow because reading will fail regardless in that case.
+    file.metadata().map(|m| m.len() as usize + 1).unwrap_or(0)
+}
+
+/// Write a slice as the entire contents of a file.
+pub fn instrumented_write<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> io::Result<()> {
+    InstrumentedFile::create(path)?.write_all(contents.as_ref())
+}
+
+/// Read the entire contents of a file into a bytes vector.
+pub fn instrumented_read<P: AsRef<Path>>(path: P) -> io::Result<Vec<u8>> {
+    let mut file = InstrumentedFile::open(path)?;
+    let mut bytes = Vec::with_capacity(initial_buffer_size(&file));
+    file.read_to_end(&mut bytes)?;
+    Ok(bytes)
+}
+
+/// Read the entire contents of a file into a string.
+pub fn instrumented_read_to_string<P: AsRef<Path>>(path: P) -> io::Result<String> {
+    let mut file = InstrumentedFile::open(path)?;
+    let mut string = String::with_capacity(initial_buffer_size(&file));
+    file.read_to_string(&mut string)?;
+    Ok(string)
+}
+
+/// Copies the contents of one file to another. This function will also
+/// copy the permission bits of the original file to the destination file.
+pub fn copy<P: AsRef<Path>, Q: AsRef<Path>>(from: P, to: Q) -> io::Result<u64> {
+    if !from.as_ref().is_file() {
+        return Err(io::Error::new(
+            ErrorKind::InvalidInput,
+            "the source path is not an existing regular file",
+        ));
+    }
+
+    let mut reader = InstrumentedFile::open(from.as_ref())?;
+    let mut writer = InstrumentedFile::create(to.as_ref())?;
+    let perm = reader.metadata()?.permissions();
+
+    let ret = io::copy(&mut reader, &mut writer)?;
```

Comment:
fsync new file and directory after copy.

Reply:
Actually this function is copied from standard implementation. Maybe instead of changing the behaviour we can add a new `copy_and_sync`?